RFC0003.md

```diff
@@ -0,0 +1,171 @@
+
+In addition, programming libraries have continued to evolve towards more of an asynchronous model where processes regularly aggregate into groups that subsequently dissolve after completing some set of operations. These new approaches would benefit from an ability to notify other processes of a desire to aggregate, and to allow the aggregation process itself to take place asynchronously.
+
+Accordingly, this RFC proposes to modify the current PMIx\_Connect[\_nb], PMIx\_Disconnect[\_nb], and their associated server-RM interfaces to accommodate the requests. It is recognized that this causes a backward-compatibility issue. However, only one library currently uses these interfaces (and they are accessed via an abstraction layer that can hide the changes) - thus, the disruption should be minimal provided these changes can be adopted and implemented before other libraries adopt them.
+
+Some additional info keys are provided for the revised operations:
+
+* PMIX\_CONNECT\_NOTIFY\_EACH: generate a local event notification using the PMIX\_PROC\_HAS\_CONNECTED event each time a process connects, and the PMIX\_PROC\_HAS\_DISCONNECTED event each time a process disconnects. The PMIx server is responsible for generating these events.
+
+* PMIX\_CONNECT\_NOTIFY\_REQ: notify each of the indicated procs that they are requested to connect using the PMIX\_CONNECT\_REQUESTED event. The PMIx server is responsible for generating this event.
+
+* PMIX\_CONNECT\_OPTIONAL: participation is optional - do not return error if procs terminate without having connected. This is the joint responsibility of the host RM (to notify if remote procs terminate early) and the PMIx server.
+
+* PMIX\_CONNECT\_EARLY\_TERM\_ALLOWED: do not generate an error if a process terminates without disconnecting. The host RM and PMIx server are responsible for adjusting completion requirements on any ongoing or future collectives for this nspace.
+
+* PMIX\_CONNECT\_XCHG\_ONLY: ensure that all participating processes have a copy of the job-level info from each nspace, but do not assign a new nspace and rank
+
+
+#### PMIx\_Connect
+Record the specified processes as "connected". Both blocking and non-blocking
+versions are provided. This means that the resource manager should treat the
+failure of any process in the specified group as a reportable event, and take
+appropriate action. Note that different resource managers may respond to
+failures in different manners.
+
+The host RM will assign a unique nspace to the resulting process group. The
+new nspace must be provided when disconnecting from the group. All procs are
+required to disconnect from the nspace prior to terminating.
+
+Note: a process can only engage in _one_ connect operation involving the identical
```

Comment:
What's the rationale for this? I would relax it to _one_ connect operation with the same namespace at the same time, or even that namespaces are unique and therefore no restrictions on simultaneity.

Reply:
We don't do a collective to assign an identifier to the connect operation, so we use the provided array of proc IDs as the identifier. However, this means that there can only be one active connect operation with that exact array of proc IDs at a time. It doesn't matter if the entire nspace is participating (i.e., a proc ID with PMIX_RANK_WILDCARD), or the array contains specific individual procs. I can adjust the language here to make it clearer if that would help.

Reply:
I updated this description - @jjhursey @abouteiller can you please review?